--- webidl_binder.py.orig	Sun Nov 15 20:06:49 2015
+++ webidl_binder.py	Thu Dec  3 00:58:52 2015
@@ -15,2 +15,4 @@
 import WebIDL
+import tempfile
+from subprocess import Popen, PIPE
 
@@ -51,2 +53,3 @@
 enums = {}
+consts = []
 
@@ -75,5 +78,16 @@
 
+mid_js += ['''
+function define(k,v) {
+  Object.defineProperty(Module, k, { value: v });
+}
+function freeze(o) {
+  Object.freeze(o);
+}
+''']
+
 def emit_constructor(name):
   global mid_js
-  mid_js += [r'''%s.prototype = %s;
+  global consts
+  consts += [name]
+  mid_js += [r'''%s.prototype = Object.create(WrapperObject.prototype);
 %s.prototype.constructor = %s;
@@ -81,4 +95,4 @@
 %s.__cache__ = {};
-Module['%s'] = %s;
-''' % (name, 'Object.create(%s.prototype)' % (implements[name][0] if implements.get(name) else 'WrapperObject'), name, name, name, name, name, name, name)]
+define('%s', %s);
+''' % (name, name, name, name, name, name, name, name)]
 
@@ -90,2 +104,10 @@
 }
+WrapperObject.prototype = {
+    free: function() {
+        destroy(this);
+    },
+    toString: function() {
+        return "[" + (this.constructor.name || "unknown") + "]";
+    }
+};
 ''']
@@ -115,3 +137,3 @@
 
-Module['NULL'] = wrapPointer(0);
+define('NULL', wrapPointer(0));
 
@@ -254,3 +276,3 @@
 def type_to_cdec(raw):
-  name = ret = type_to_c(raw.type.name, non_pointing=True)
+  name = ret = type_to_c(raw.type.name, non_pointing=False)
   if raw.getExtendedAttribute('Const'): ret = 'const ' + ret
@@ -445,5 +467,5 @@
     %sEM_ASM_%s({
-      var self = Module['getCache'](Module['%s'])[$0];
+      var self = getCache(%s)[$0];
       if (!self.hasOwnProperty('%s')) throw 'a JSImplementation must implement all functions, you forgot %s::%s.';
-      %sself['%s'](%s)%s;
+      %sself.%s(%s)%s;
     }, (int)this%s);
@@ -455,3 +477,3 @@
           func_name,
-          ','.join(['$%d' % i for i in range(1, max_args + 1)]),
+          ','.join(['wrapPointer($%d, %s)' % (i, raw[i-1].type.name) for i in range(1, max_args + 1)]),
           return_postfix,
@@ -623,16 +645,15 @@
 deferred_js = []
+tocompile_c = []
 
 for name, enum in enums.iteritems():
-  mid_c += ['\n// ' + name + '\n']
-  deferred_js += ['\n', '// ' + name + '\n']
+  tocompile_c += ['\n', '// ' + name + '\n']
   for value in enum.values():
     function_id = "%s_%s" % (name, value.split('::')[-1])
-    mid_c += [r'''%s EMSCRIPTEN_KEEPALIVE emscripten_enum_%s() {
-  return %s;
-}
-''' % (name, function_id, value)]
     symbols = value.split('::')
+    while len(symbols) > 2:
+      symbols.pop(0);
+
     if len(symbols) == 1:
       identifier = symbols[0]
-      deferred_js += ["Module['%s'] = _emscripten_enum_%s();\n" % (identifier, function_id)]
+      tocompile_c += ["printf(\"Module.%s = %%d;\\n\", %s);\n" % (identifier, value)]
     elif len(symbols) == 2:
@@ -641,7 +662,7 @@
         # namespace is a class
-        deferred_js += ["Module['%s']['%s'] = _emscripten_enum_%s();\n" % \
-                  (namespace, identifier, function_id)]
+        tocompile_c += ["printf(\"%s.%s = %%d;\\n\", %s);\n" % \
+                  (namespace, identifier, value)]
       else:
         # namespace is a namespace, so the enums get collapsed into the top level namespace.
-        deferred_js += ["Module['%s'] = _emscripten_enum_%s();\n" % (identifier, function_id)]
+        tocompile_c += ["printf(\"Module.%s = %%d;\\n\", %s);\n" % (identifier, value)]
     else:
@@ -649,11 +670,42 @@
 
+tocompile_c = '''
+#include <megaapi.h>
+#include <stdio.h>
+
+int main() {
+    %s
+    return 0;
+}
+''' % '\n    '.join(tocompile_c)
+
+temp = tempfile.mkstemp('.cpp');
+
+c = open(temp[1], 'w+')
+c.write(tocompile_c)
+c.close()
+
+process = Popen(["g++", temp[1], "-o", temp[1] + ".exe", "-Iinclude"], stdout=PIPE)
+(output, err) = process.communicate()
+exit_code = process.wait()
+
+if (exit_code):
+    sys.exit(1)
+
+process = Popen([temp[1] + ".exe"], stdout=PIPE)
+(output, err) = process.communicate()
+exit_code = process.wait()
+
+if (exit_code):
+    sys.exit(1)
+
+deferred_js += [output]
+
+consts = set(consts)
+for x in consts:
+    deferred_js += ["freeze(%s);" % x]
+
 mid_c += ['\n}\n\n']
 mid_js += ['''
-(function() {
-  function setupEnums() {
-    %s
-  }
-  if (Module['calledRun']) setupEnums();
-  else addOnPreMain(setupEnums);
-})();
+// Setup enums
+%s
 ''' % '\n    '.join(deferred_js)]
